### ISSUE TITLE
Make Cheerio optional

### ISSUE THREAD

**Issue by khirakawa** [airbnb/enzyme#1310]
Importing `enzyme` in an react native application results in this error at build time:

```
error: bundling failed: UnableToResolveError: Unable to resolve module `events` from `some/path/to/node_modules/enzyme/node_modules/htmlparser2/lib/Parser.js`: Module does not exist in the module map
```

This seems to be coming from the various `cheerio` imports in the `enzyme` package.  It would be great if we made `cheerio` optional or perhaps lazy load `cheerio` inside the various `render` functions that use it:

```javascript
render() {
   const html = this.html();
   const cheerio = require('cheerio');
   return html === null ? cheerio() : cheerio.load('')(html);
}
```

**Comment by ljharb** [comment]
I'm not sure what the issue would be; what is "the module map"? `require` should Just Work with everything that's on disk.

**Comment by khirakawa** [comment]
Apologies, I was able to fix this by `npm install --save-dev events` 👍

**Comment by ljharb** [comment]
Always make sure `npm ls` exits zero, otherwise nothing can be expected to work :-)

**Comment by khirakawa** [comment]
Looks like I spoke too soon.  After installing `events`, `buffer`, and `stream`, (and making sure `npm ls` exists with zero), I am getting the following error:

`o.prototype is undefined`

Digging through the callstack, 

1. Error is thrown [here](https://github.com/isaacs/inherits/blob/master/inherits_browser.js#L5) in the `inherits` module
2. `inherits` is called in [`parse5/lib/parser/parser_stream.js`](https://github.com/inikulin/parse5/blob/master/lib/parser/parser_stream.js#L27).  When the error is thrown, `WritableStream` is `undefined`
2. It's `undefined` because on [this line](https://github.com/inikulin/parse5/blob/master/lib/parser/parser_stream.js#L3) `require('stream')` doesn't export a `Writable` object.  See [stream](https://github.com/juliangruber/stream/blob/master/index.js)

Unfortunately, this puts me back to my original problem where I can't load `enzyme`.  As a workaround, I can fork `enzyme`, remove calls to `cheerio` or stub it via `babel-plugin-module-rewrite`, but I'd rather not do that if possible.

**Comment by ljharb** [comment]
I'm a bit confused; the problem you're describing is a dependency graph problem, which either 100% or 0% of enzyme users would be running into.

What version of `node` and `npm` are you using? Is this only occurring when browserifying or webpacking enzyme? Do your tests pass in `node` itself, which is the primary use case of enzyme?

**Comment by khirakawa** [comment]
Apologies for the confusion.  We're building an integration test framework that runs as a react-native application on an actual device (non-node environment).  For this, we're using `mocha` and `enzyme` with our custom enzyme adapter, and hence why we are importing `enzyme` inside a react-native application.  It works great, as long as I remove `cheerio` imports in `enzyme`.

This is obviously _not_ an intended use case for `enzyme`, but it would be great if we can come to some sort of solution.  We do have a workaround (see above), but it requires forking `enzyme`, which we'd like to avoid.

To answer your questions:

> What version of `node` and `npm` are you using?

- node: 8.6.0
- npm: 5.3.0

> Is this only occurring when browserifying or webpacking enzyme

This occurs only when used with the react native packager.

> Do your tests pass in node itself

If:

1. We run in `node` instead of on an actual device
2. Stub out native dependencies
3. Use `enzyme-adapter-react-16` instead of our custom adapter

Then, yes, it does run 👍 

Again though, our intended goal isn't to run on node, but on an actual device.

**Comment by ljharb** [comment]
Since you can stub out native dependencies, can you not stub out `cheerio` in the RNP, and simply not use `render`?

**Comment by khirakawa** [comment]
Good point.  I did initially try stubbing `cheerio` using the stubbing mechanism used in [react-native-mock](https://github.com/RealOrangeOne/react-native-mock), but it turns out `react-native-mock` is a run-time mock that relies on the require cache being exposed.  Unfortunately, the error thrown when importing `enzyme` in RN is a build-time issue.  The require cache isn't exposed either.

However, I recently stumbled upon https://github.com/philikon/ReactNativify, which aims to address this exact issue.  It turns out `babel-plugin-rewrite-require` works:

In `.babelrc`:

```
{
  "plugins": [
    ["babel-plugin-rewrite-require", {
      "aliases": {
        "cheerio": "my-framework/cheerio-stub"
      }
    }]
  ]
}
```

Where `my-framework/cheerio-stub` is just `module.exports = {};`

This is much better than having to fork `enzyme` 👍 .   If there was a browserified `cheerio`, we could use that instead as well.

It would still be great if `cheerio` was optional so that all consumers of our framework didn't need to install this plugin + modify their `.babelrc`, but this will be sufficient for us.  Actually now that I think about it, I think what I'm really requesting for is for `enzyme` to support browserify, webpack, _and_ RN's metro bundler. 😄 

Thanks for your help @ljharb!
